crates/engine/tree/src/tree/cached_state.rs

```diff
@@ -0,0 +1,288 @@
+    }
+
+    /// Returns a new zeroed-out instance of [`CachedStateMetrics`].
+    pub(crate) fn zeroed() -> Self {
+        let zeroed = Self::default();
+        zeroed.reset();
+        zeroed
+    }
+}
+
+impl<S: AccountReader> AccountReader for CachedStateProvider<S> {
+    fn basic_account(&self, address: &Address) -> ProviderResult<Option<Account>> {
+        if let Some(res) = self.caches.account_cache.get(address) {
+            self.metrics.account_cache_hits.increment(1);
+            return Ok(res)
+        }
+
+        self.metrics.account_cache_misses.increment(1);
+
+        let res = self.state_provider.basic_account(address)?;
+        self.caches.account_cache.insert(*address, res);
+        Ok(res)
+    }
+}
+
+impl<S: StateProvider> StateProvider for CachedStateProvider<S> {
+    fn storage(
+        &self,
+        account: Address,
+        storage_key: StorageKey,
+    ) -> ProviderResult<Option<StorageValue>> {
+        if let Some(res) = self.caches.storage_cache.get(&(account, storage_key)) {
+            self.metrics.storage_cache_hits.increment(1);
+            return Ok(res)
+        }
+
+        self.metrics.storage_cache_misses.increment(1);
+
+        let final_res = self.state_provider.storage(account, storage_key)?;
+        self.caches.storage_cache.insert((account, storage_key), final_res);
+        Ok(final_res)
+    }
+
+    fn bytecode_by_hash(&self, code_hash: &B256) -> ProviderResult<Option<Bytecode>> {
+        if let Some(res) = self.caches.code_cache.get(code_hash) {
+            self.metrics.code_cache_hits.increment(1);
+            return Ok(res)
+        }
+
+        self.metrics.code_cache_misses.increment(1);
+
+        let final_res = self.state_provider.bytecode_by_hash(code_hash)?;
+        self.caches.code_cache.insert(*code_hash, final_res.clone());
+        Ok(final_res)
+    }
+}
+
+impl<S: StateRootProvider> StateRootProvider for CachedStateProvider<S> {
+    fn state_root(&self, hashed_state: HashedPostState) -> ProviderResult<B256> {
+        self.state_provider.state_root(hashed_state)
+    }
+
+    fn state_root_from_nodes(&self, input: TrieInput) -> ProviderResult<B256> {
+        self.state_provider.state_root_from_nodes(input)
+    }
+
+    fn state_root_from_nodes_with_updates(
+        &self,
+        input: TrieInput,
+    ) -> ProviderResult<(B256, TrieUpdates)> {
+        self.state_provider.state_root_from_nodes_with_updates(input)
+    }
+
+    fn state_root_with_updates(
+        &self,
+        hashed_state: HashedPostState,
+    ) -> ProviderResult<(B256, TrieUpdates)> {
+        self.state_provider.state_root_with_updates(hashed_state)
+    }
+}
+
+impl<S: StateProofProvider> StateProofProvider for CachedStateProvider<S> {
+    fn proof(
+        &self,
+        input: TrieInput,
+        address: Address,
+        slots: &[B256],
+    ) -> ProviderResult<AccountProof> {
+        self.state_provider.proof(input, address, slots)
+    }
+
+    fn multiproof(
```

Comment:
What's the plan for this? Also these caches should probably work across blocks too?

Reply:
rather than caching proofs right now, we'll send messages to the state root task, which will use a different provider to fetch + cache proofs itself for each block

We should start with intra-block caching, but for cross-block caching, we would need to apply state updates to the cache, instead of invalidating each block. Maybe we could also incorporate trie updates as well

Reply:
Yes exactly